kernel/src/driver.rs

```diff
@@ -33,82 +39,162 @@
 //!
-//! # The `yield` System-call
+//! # The `yield` system call class
+//!
+//! While drivers do not handle `yield` system calls, it is important
+//! to understand them and how they interacts with `subscribe`, which
+//! registers callback functions with the kernel. When a process calls
+//! a `yield` system call, the kernel checks if there are any pending
+//! callbacks for the process. If there are pending callbacks, it
+//! pushes one callback onto the process stack. If there are no
+//! pending callbacks, `yield-wait` will cause the process to sleep
+//! until a callback is trigered, while `yield-no-wait` returns
+//! immediately.
+//!
+//! # Method result types
 //!
-//! While drivers do not handle the `yield` system call, it is important to
-//! understand its function and how it interacts with `subscribe`.
+//! Each driver method has a limited set of valid return types. Every
+//! method has a single return type corresponding to success and a
+//! single return type corresponding to failure. For the `subscribe`
+//! and `allow` system calls, these return types are the same for
+//! every instance of those calls. Each instance of the `command`
+//! system call, however, has its own specified return types. A
+//! command that requests a timestamp, for example, might return a
+//! 32-bit number on success and an error code on failure, while a
+//! command that requests time of day in microsecond granularity might
+//! return a 64-bit number and a 32-bit timezone encoding on success,
+//! and an error code on failure.
+//!
+//! These result types are represented as safe Rust types. The core
+//! kernel (the scheduler and syscall dispatcher) is responsible for
+//! encoding these types into the Tock system call ABI specification.
 
 use crate::callback::{AppId, Callback};
-use crate::mem::{AppSlice, Shared};
+use crate::errorcode::ErrorCode;
+use crate::mem::{ReadOnlyAppSlice, ReadWriteAppSlice};
+use crate::process;
 use crate::returncode::ReturnCode;
+use crate::syscall::GenericSyscallReturnValue;
 
-/// `Driver`s implement the three driver-specific system calls: `subscribe`,
-/// `command` and `allow`.
+/// Possible return values of a `command` driver method
+///
+/// This is just a wrapper around
+/// [`GenericSyscallReturnValue`](GenericSyscallReturnValue) since a
+/// `command` driver method may only return primitve integer types as
+/// payload.
 ///
-/// See [the module level documentation](index.html) for an overview of how
-/// system calls are assigned to drivers.
+/// It is important for this wrapper to only be constructable over
+/// variants of
+/// [`GenericSyscallReturnValue`](GenericSyscallReturnValue) that are
+/// deemed safe for a capsule to construct and return to an
+/// application (e.g. not
+/// [`SubscribeSuccess`](crate::syscall::GenericSyscallReturnValue::SubscribeSuccess)).
+/// This means that the inner value **must** remain private.
+pub struct CommandResult(GenericSyscallReturnValue);
```

Comment:
Should these names parallel each other? For example:

```suggestion
pub struct CommandReturnValue(SyscallReturnValue);
```

Using `Result` is confusing, since it doesn't look like a Rust `Result`. I think `GenericSyscallReturnValue` is a little clumsy, since there isn't a `SpecificSyscallReturnValue`, and the type does in fact include all possible syscall return variants.

Reply:
For what it's worth, I've been using `CommandReturn` in `libtock-rs`.

Reply:
CommandReturn sounds good. Better to be consistent with libtock-rs.